src/Features/Core/Portable/SignatureHelp/SignatureHelpServiceWithProviders.cs

```diff
@@ -0,0 +1,285 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Linq;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.SignatureHelp
+{
+    /// <summary>
+    /// A subtype of <see cref="SignatureHelpService"/> that aggregates signatures from one or more <see cref="SignatureHelpProvider"/>s.
+    /// </summary>
+    internal abstract class SignatureHelpServiceWithProviders : SignatureHelpService
+    {
+        private readonly Workspace _workspace;
+        private readonly string _language;
+
+        public SignatureHelpServiceWithProviders(Workspace workspace, string language)
+        {
+            _workspace = workspace;
+            _language = language;
+        }
+
+        protected virtual ImmutableArray<SignatureHelpProvider> GetBuiltInProviders()
+        {
+            return ImmutableArray<SignatureHelpProvider>.Empty;
+        }
+
+        private ImmutableArray<SignatureHelpProvider> _importedProviders;
+
+        private IEnumerable<SignatureHelpProvider> GetImportedProviders()
+        {
+            if (_importedProviders.IsDefault)
+            {
+                var mefExporter = (IMefHostExportProvider)_workspace.Services.HostServices;
+
+                var providers = ExtensionOrderer.Order(
+                        mefExporter.GetExports<SignatureHelpProvider, SignatureHelpProviderMetadata>()
+                        .Where(lz => lz.Metadata.Language == _language)
+                        ).Select(lz => lz.Value).ToImmutableArray();
+
+                ImmutableInterlocked.InterlockedCompareExchange(ref _importedProviders, providers, default(ImmutableArray<SignatureHelpProvider>));
+            }
+
+            return _importedProviders;
+        }
+
+        private ImmutableArray<SignatureHelpProvider> _testProviders = ImmutableArray<SignatureHelpProvider>.Empty;
+        private bool _testAugmentBuiltInProviders;
+
+        internal void SetTestProviders(IEnumerable<SignatureHelpProvider> testProviders, bool augmentBuiltInProviders = false)
```

Comment:
given that we can do this, is tehre a need to have you special test service in tests?  Can we not just call SetTestProviders with the single provider we cared about?